File: src/graph_algorithms/breadth_first_search.py
from collections import deque

def breadth_first_search(graph, start):
    """
    Perform Breadth First Search (BFS) on a graph.

    Parameters:
        graph (dict): A dictionary representing the adjacency list of the graph.
                      Each key is a vertex, and the value is a list of neighbors.
        start (any): The starting vertex for the BFS.

    Returns:
        list: A list of vertices in the order they are visited.
    """
    visited = set()
    queue = deque([start])
    result = []

    while queue:
        vertex = queue.popleft()

        if vertex not in visited:
            visited.add(vertex)
            result.append(vertex)

            # Add all unvisited neighbors to the queue
            for neighbor in graph[vertex]:
                if neighbor not in visited:
                    queue.append(neighbor)

    return result

File: src-tauri/src/Algorithms/graph_algorithms/breadth_first_search.rs
use std::collections::VecDeque;

pub fn bfs(graph: &Vec<Vec<usize>>, start: usize) -> Vec<usize> {
    let mut visited = vec![false; graph.len()];
    let mut queue = VecDeque::new();
    let mut order = Vec::new();

    visited[start] = true;
    queue.push_back(start);

    while let Some(node) = queue.pop_front() {
        order.push(node);

        for &neighbor in &graph[node] {
            if !visited[neighbor] {
                visited[neighbor] = true;
                queue.push_back(neighbor);
            }
        }
    }

    order
}

fn main() {
    let graph = vec![vec![1, 2], vec![2], vec![0, 3], vec![3]];

    let start = 2;
    let bfs_order = bfs(&graph, start);

    println!("BFS starting from node {}: {:?}", start, bfs_order);
}
